grades/tasks_test.py

```diff
@@ -0,0 +1,173 @@
+
+    def test_freeze_course_run_final_grades_3(self):
+        """
+        Test for the test_freeze_course_run_final_grades
+        task in case there are no more users to be processed
+        """
+        run_grade_info_qset = CourseRunGradingStatus.objects.filter(course_run=self.course_run2)
+        assert run_grade_info_qset.exists() is False
+        tasks.freeze_course_run_final_grades.delay(self.course_run2)
+        assert run_grade_info_qset.exists() is True
+        info_run = run_grade_info_qset.first()
+        assert info_run.course_run == self.course_run2
+        assert info_run.status == FinalGradeStatus.COMPLETE
+
+    @patch('grades.api.freeze_user_final_grade', autospec=True)
+    def test_freeze_course_run_final_grades_4(self, freeze_single_user):
+        """
+        Test for the test_freeze_course_run_final_grades
+        task in case there are users to be processed
+
+        NOTE: In this test it is mocked a function in a subtask and not the subtask.
+        The reason is because mocking subtasks inside celery groups creates problems.
+        """
+        # first call
+        run_grade_info_qset = CourseRunGradingStatus.objects.filter(course_run=self.course_run1)
+        assert run_grade_info_qset.exists() is False
+        tasks.freeze_course_run_final_grades.delay(self.course_run1)
+        assert run_grade_info_qset.exists() is True
+        info_run = run_grade_info_qset.first()
+        assert info_run.course_run == self.course_run1
+        assert info_run.status == FinalGradeStatus.PENDING
+        cache_id = tasks.CACHE_ID_BASE_STR.format(self.course_run1.edx_course_key)
+        cached_celery_task_id_1 = cache.get(cache_id)
+        assert cached_celery_task_id_1 is not None
+        assert freeze_single_user.call_count == len(self.users)
+        for user in self.users:
+            freeze_single_user.assert_any_call(user, self.course_run1)
+
+        # simulate successful freeze for most users
+        successful_users = self.users[5:]
+        for user in successful_users:
+            FinalGrade.objects.create(
+                user=user,
+                grade=0.6,
+                passed=True,
+                course_run=self.course_run1,
+                status=FinalGradeStatus.COMPLETE
+            )
+
+        # new call will process all the remaining users without changing the status of the course run
+        freeze_single_user.reset_mock()
+        tasks.freeze_course_run_final_grades.delay(self.course_run1)
+        info_run.refresh_from_db()
+        assert info_run.status == FinalGradeStatus.PENDING
+        cached_celery_task_id_2 = cache.get(cache_id)
+        assert cached_celery_task_id_2 is not None
+        assert cached_celery_task_id_2 != cached_celery_task_id_1
+        remaining_users = self.users[:5]
+        assert freeze_single_user.call_count == len(remaining_users)
+        for user in remaining_users:
+            freeze_single_user.assert_any_call(user, self.course_run1)
```

Comment:
Can you also assert that `freeze_single_user` was not called again for `successful_users`?